text/0000-tuple-struct-self-ctor.md

```diff
@@ -0,0 +1,198 @@
+larger wherefore the language becomes more complex.
+
+There are good reasons why `Self { f0: v0, f1: v1, .. }` is allowed.
+Chiefly among those is that it becomes easier to refactor the code when
+one wants to rename type names. Another important reason is that only
+having to keep `Self` in mind means that a developer does not need to
+keep the type name fresh in their working memory. This is beneficial for
+users with shorter working memory such as the author of this RFC.
+
+Since `Self { f0: v0, .. }` is well motivated, those benefits and motivations
+will also extend to tuple structs. Eliminating this discrepancy between tuple
+structs and those with named fields will therefore have all the benefits
+associated with this feature for structs with named fields.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Basic concept
+
+For structs with named fields such as:
+
+```rust
+struct Person {
+    name: String,
+    ssn: usize,
+    age: usize
+}
+```
+
+You may use the syntax `Self { field0: value0, .. }` as seen below
+instead of writing `TypeName { field0: value0, .. }`:
+
+```rust
+impl Person {
+    /// Make a newborn person.
+    fn newborn(name: String, ssn: usize) -> Self {
+        Self { name, ssn, age: 0 }
+    }
+}
+```
+
+## Through type aliases
+
+This ability does not extend to tuple structs however in current Rust but will
+with this RFC. To continue on with the previous example, you can now also write:
+
+```rust
+struct Person(String, usize, usize);
+
+impl Person {
+    /// Make a newborn person.
+    fn newborn(name: String, ssn: usize) -> Self {
+        Self(name, ssn, 0)
+    }
+}
+```
+
+As with structs with named fields, you may also use `Self` when
+you are `impl`ing on a type alias of a struct as seen here:
+
+```rust
+struct FooBar(u8);
+
+type BarFoo = FooBar;
+
+impl Default for BarFoo {
+    fn default() -> Self {
+        Self(42) // <-- Not allowed before.
+    }
+}
+```
+
+## Teaching the contents
+
+This RFC should not require additional effort other than spreading the
+news that this now is possible as well as the reference. The changes are
+seen as intuitive enough that it supports what the user already assumes
+should work and will probably try at some point.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Rust (now) allows usage of `Self(v0, v1, ..)` inside inherent
+and trait `impl`s of tuple structs, either when mentioning the
+tuple struct directly in the `impl` header, or via a type alias.
+
+## Desugaring
+
+When the compiler encounters the following syntactic form specified in `EBNF`:
+
+```ebnf
+SelfTupleApply ::= "Self" "(" ExprList ")" ;
+ExprList ::= Expr "," Values | Expr | "" ;
+```
+
+the compiler will desugar the application by substituting `Self(v0, v1, ..)`
+for `Self { 0: v0, 1: v1, .. }` and then continue on from there. The compiler
```

Comment:
This desugaring is not entirely equivalent to what constructors tuple struct scan do. In
```rust
struct S(u8);
```
`S` in value namespace is a constructor function and it can be used in all contexts where a `fn(u8) -> S` is accepted, not only direct calls `S(arg)`.
```rust
(S)(0u8); // OK
opt_u8.map(S); // OK
(Self)(0u8); // Not OK, not a direct call
opt_u8.map(Self); // Not OK, not a direct call
```

Ideally, `Self` should be a first-class "function alias" and refer to the constructor function (with appropriate generic parameter substitutions).

Reply:
>`Self` should be a first-class "function alias"

Or a "const alias" for unit structs. "Value alias", in general.

Reply:
That's strictly more powerful I guess, so I could change the desugaring to go in this direction if we want to support that. Do we want to support that? Why / why not?